### PR TITLE
Added test to density that will run maximum capacity pods on nodes

### PR DESCRIPTION
**Pull request by rrati**
Added a test to the Density Suite that will load the kubelets with their maximum capacity number of pods

### CODE REVIEW

**Comment by gmarek** [comment]
We're running kubemark tests of this kind. Do you want to run this on real nodes? If so it's probably a @kubernetes/goog-node thing.

**Comment by timothysc** [comment]
@gmarek we want to run it against a real cluster, we could denote the test to be skipped by default if needed.

**Comment by gmarek** [comment]
Also we're running maximum number of Pods in SchedulerPredicates test.

**Comment by rrati** [comment]
There are some things in SchedulerPredicates that Density should probably be aware of, like cluster pods, but SchedulerPredicates doesn't have the metrics that density produces and looks to be more focused on the scheduler side.

**Comment by yujuhong** [comment]
Does the test consider the cluster add-on pods? If not, it'd fail.

File: test/e2e/density.go [hunk 1]
```diff
@@ -562,6 +587,45 @@ var _ = framework.KubeDescribe("Density", func() {
 			}
 		})
 	}
+
+	It("should allow running maximum capacity pods on nodes", func() {
+		// Calculate total number of pods from each node's max-pod
+		totalPods = 0
+		for _, n := range nodes.Items {
```
Comment (gmarek) [inline test/e2e/density.go]: This won't work, as you need to take into account Pods already running in the cluster (e.g. KubeProxy on each node). Take a look at SchedulerPredicates test.

**Comment by rrati** [comment]
None of the density tests currently take cluster pods into account, but that may not matter with the others.  I'll update this one to take cluster pods into account

File: test/e2e/density.go [hunk 2]
```diff
@@ -562,6 +587,45 @@ var _ = framework.KubeDescribe("Density", func() {
 			}
 		})
 	}
+
+	It("should allow running maximum capacity pods on nodes", func() {
```
Comment (gmarek) [inline test/e2e/density.go]: Prefix the name with `[Feature:ManualPerformance]`

**Comment by vishh** [comment]
Question: Why not add this test to the node e2e suite? If there are no cluster level functionalities being tested, this test can go under `test/e2e_node`...

**Comment by rrati** [comment]
The desire is to be able to run this against a real cluster.  Get metrics/information for how nodes/system performs when loaded with maximum pods.  Next step after this would be to use pods that do activity rather than just pause containers.

File: test/e2e/density.go (hunk 2 above)
Comment (rrati) [inline test/e2e/density.go]: Done

**Comment by wojtek-t** [comment]
@rrati - I completely agree with @vishh here. This should be part of node e2e suite.

If you only motivation is having some metrics, note that those are pretty well extracted and you can just use them adding two calls in your test:
- ResetMetrics on the beginning of the tst
- HighLatencyRequests at the end
See:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/metrics_util.go


Also, can you please clarify what you are trying to achieve with this test?

File: test/e2e/density.go (hunk 1 above)
Comment (rrati) [inline test/e2e/density.go]: Done

**Comment by rrati** [comment]
@wojtek-t @vishh The motivation is to be able to run a cluster at max-pods for each node, get metrics, etc.  That seems similar in use case to the tests in the density suite, vs what I see in e2e_node.

**Comment by timothysc** [comment]
We need a test that can execute max-pods in a heterogeneous env and collect metrics across a cluster.  We will leverage pods-core for this test.

**Comment by vishh** [comment]
What are the metrics of interest? A test like this would be useful to catch
system issues under load. If it is just metrics though, doesn't the
existing kubemark tests and node e2e tests provide necessary metrics?

**Comment by timothysc** [comment]
Density can at scale can determine the scheduling throughput at max capacity which is highly useful.

**Comment by gmarek** [comment]
My (another 2c) here. It's quite hard to schedule max-pods * number of nodes - see all the problems that we had with SchedulerPredicates test. This makes me strongly against purely 'max-pods' density test - it just won't work reliably enough.

Is the main reason for this PR to have a test that will grow together with `max-pods` value? If so may I suggest doing it as an ordinary Density test with value `max-pods - 3` (or sth)? It won't work for small clusters, but I'm not sure we care.

**Comment by rrati** [comment]
@gmarek What issues have you had scheduling to max capacity on nodes?  Scheduler or kubelet issues?

**Comment by gmarek** [comment]
Kubelet issues - it's not easy to make sure that everything that should be running is running. @dchen1107

**Comment by goltermann** [comment]
@rrati, can you rebase and get this merged?  we're cutting the 1.3 branch for beta on Friday June 10

**Comment by timothysc** [comment]
Folks - 

We would like to get this test in for our performance analysis against 1.4, and would be happy to label as [Disruptive], where we can execute manually on our side as part of our test plans.  This test primarily serves as a benchmark to fill a cluster to maximum capacity, and is not intended to be a node e2e-test. 

Thoughts?  @gmarek

**Comment by gmarek** [comment]
@timothysc - sorry for the late answer, I was on vacation. Can you write what do you want to measure? Time for full saturation or some derivative of it (scheduler throughput around the point of full capacity, etc.), API call latency for full cluster with N nodes, something else?

I'm asking because there might be some other way to do this, as reaching the limit is known to be flaky (some system pods might be restarting, image-puller might be running or just finished, etc.). I think that running "close to full" (i.e. full - 2*Nodes - all known pods or something) would be way less flaky, if that's enough for you.

**Comment by timothysc** [comment]
We want to plot the curve to full capacity, and measure system overhead and latencies in the system.

**Comment by gmarek** [comment]
By full you mean full full, or close to full is enough? I really don't think that running at absolute max is a good idea, but if want to maintain this and put resources behind making it not for flaky then I'm OK with it.

**Comment by timothysc** [comment]
full full, we explicitly want to find & fix where it breaks.

**Comment by gmarek** [comment]
@timothysc - OK, I'll review this CL, but I count on you for maintaining the test and making it not-flaky.

File: test/e2e/density.go [hunk 3]
```diff
@@ -166,6 +176,152 @@ func logPodStartupStatus(c *client.Client, expectedPods int, ns string, observed
 	}
 }
 
+func runDensityTest(dtc DensityTestConfig, startupTime *time.Duration) {
```
Comment (gmarek) [inline test/e2e/density.go]: Why do you need a pointer to the startupTime?

File: test/e2e/framework/util.go [hunk 4]
```diff
@@ -4425,3 +4426,64 @@ func retryCmd(command string, args ...string) (string, string, error) {
...
+					notScheduledPods = append(notScheduledPods, pod)
+				}
+			}
+		}
+	}
+	return
+}
+
+// WaitForStableCluster waits until all existing pods are scheduled and returns their amount.
+func WaitForStableCluster(c *client.Client, masterNodes sets.String) int {
```
Comment (gmarek) [inline test/e2e/framework/util.go]: Please update this function and remove it from scheduler_predicates.

Comment (gmarek) [inline test/e2e/framework/util.go]: Remove from scheduler_predicates as well.

File: test/e2e/framework/util.go [hunk 5]
```diff
@@ -4425,3 +4426,64 @@ func retryCmd(command string, args ...string) (string, string, error) {
...
+
+// GetMasterAndWorkerNodesOrDie will return a list schedulable master and works nodes
+func GetMasterAndWorkerNodesOrDie(c *client.Client) (sets.String, *api.NodeList) {
+	nodes := &api.NodeList{}
+	masters := sets.NewString()
+	all := GetReadySchedulableNodesOrDie(c)
+	for _, n := range all.Items {
+		if system.IsMasterNode(&n) {
+			masters.Insert(n.Name)
+		} else {
```
Comment (gmarek) [inline test/e2e/framework/util.go]: As we're updating it either way it's probably good to have this return schedulable Nodes only.

File: test/e2e/density.go [hunk 6]
```diff
@@ -631,6 +659,44 @@ var _ = framework.KubeDescribe("Density", func() {
...
+		for i := 0; i < rcCnt; i++ {
+			RCName = "density" + strconv.Itoa(totalPods) + "-" + strconv.Itoa(i) + "-" + uuid
+			RCConfigs[i] = framework.RCConfig{Client: c,
+				Image:                "gcr.io/google_containers/pause-amd64:3.0",
+				Name:                 RCName,
+				Namespace:            ns,
+				Labels:               map[string]string{"type": "densityPod"},
+				PollInterval:         10 * time.Second,
+				PodStatusFile:        fileHndl,
+				Replicas:             (totalPods + rcCnt - 1) / rcCnt,
```
Comment (gmarek) [inline test/e2e/density.go]: It won't saturate the cluster unless `totalPods + rcCnt - 1` is for some reason divisible by `rcCnt` (it's OK for 1, but please make it work for other values as well).

File: test/e2e/framework/util.go (hunk 4 above)
Comment (rrati) [inline test/e2e/framework/util.go]: This was originally removed.  Must have been added back from a rebase.

Comment (gmarek) [inline test/e2e/framework/util.go]: Make sure that the logic didn't change then.

File: test/e2e/framework/util.go (hunk 5 above)
Comment (rrati) [inline test/e2e/framework/util.go]: If we're not going to have this function return masters anymore, does it make sense to modify GetMasterHost to be more comprehensive?  This returns all masters in the cluster whereas GetMasterHost always returns 1.  If there is a desire for testing against HA, seems we would want a method to get all masters.

File: test/e2e/density.go (hunk 3 above)
Comment (rrati) [inline test/e2e/density.go]: The function determines the startupTime which is used in AfterEach..  It is probably better to have that as a return from the function though.

File: test/e2e/framework/util.go (hunk 5 above)
Comment (gmarek) [inline test/e2e/framework/util.go]: Oh, sorry, I wasn't clear. We certainly want it to return master nodes, but we probably don't want to return all other nodes, but only schedulable ones.

File: test/e2e/density.go (hunk 3 above)
Comment (gmarek) [inline test/e2e/density.go]: Right:)

Comment (rrati) [inline test/e2e/density.go]: Done

File: test/e2e/framework/util.go (hunk 4 above)
Comment (rrati) [inline test/e2e/framework/util.go]: Done

File: test/e2e/framework/util.go (hunk 5 above)
Comment (rrati) [inline test/e2e/framework/util.go]: It should already be doing that?  It only iterates over nodes that are schedulable, so the masters/nodes returned should all be schedulable.  Am I missing something?

File: test/e2e/density.go (hunk 6 above)
Comment (rrati) [inline test/e2e/density.go]: Done

File: test/e2e/framework/util.go (hunk 5 above)
Comment (gmarek) [inline test/e2e/framework/util.go]: Oh, sorry - your PR is based on one that got reverted (one that did this "GetSchedulableNodesOrDie" thing here - this completely breaks this test, as master isn't schedulable.

Comment (rrati) [inline test/e2e/framework/util.go]: Done

File: test/e2e/density.go [hunk 7]
```diff
@@ -167,6 +177,153 @@ func logPodStartupStatus(c *client.Client, expectedPods int, ns string, observed
 	}
 }
 
+func runDensityTest(dtc DensityTestConfig) time.Duration {
```
Comment (gmarek) [inline test/e2e/density.go]: Add a comment what the return value is.

File: test/e2e/density.go [hunk 8]
```diff
@@ -632,6 +661,48 @@ var _ = framework.KubeDescribe("Density", func() {
...
+
+	// Calculate total number of pods from each node's max-pod
+	It("[Feature:ManualPerformance] should allow running maximum capacity pods on nodes", func() {
+		totalPods = 0
+		for _, n := range nodes.Items {
+			totalPods += int(n.Status.Capacity.Pods().Value())
+		}
+		totalPods -= framework.WaitForStableCluster(c, masters)
+
+		fileHndl, err := os.Create(fmt.Sprintf(framework.TestContext.OutputDir+"/%s/pod_states.csv", uuid))
```
Comment (gmarek) [inline test/e2e/density.go]: You probably want to have a flag or env var that enables/disables writing output files (when we had this kind of behavior previously it ended in a number of disk full problems).

**Comment by gmarek** [comment]
Two nits, LGTM otherwise. You can apply the second nit or ignore it and apply LGTM label yourself after you add a comment to the `runDensityTest` function.

File: test/e2e/density.go (hunk 7 above)
Comment (rrati) [inline test/e2e/density.go]: Done

File: test/e2e/density.go (hunk 8 above)
Comment (rrati) [inline test/e2e/density.go]: The disk usage should end up being similar to other density tests and they haven been small text files that haven't caused a problem (yet).  This data file is pretty important for our performance evaluations, so for the time being I think I would like to keep it like the other density tests.  If this file causes disk space issues something is seriously wrong (like the test timeout doesn't work).

**Comment by timothysc** [comment]
@jeremyeder more additions to test plans.  

We need a way to automate this.

**Pull request merged**